diff --git a/src/include/ifxos_version.h b/src/include/ifxos_version.h
--- a/src/include/ifxos_version.h
+++ b/src/include/ifxos_version.h
@@ -105,7 +105,7 @@
 /** IFXOS version - feature */
 #define IFXOS_VERSION_FEATURE                 5
 /** IFXOS version, build number - step */
-#define IFXOS_VERSION_STEP                    16
+#define IFXOS_VERSION_STEP                    17
 
 /** IFXOS version as number */
 #define IFXOS_VER_NUMBER \
