disable dhrystone command
--- a/include/configs/ifx_mips.h
+++ b/include/configs/ifx_mips.h
@@ -72,7 +72,9 @@
 
 #define CONFIG_COMMANDS_YES	(CONFIG_CMD_DFL 	| \
 				 CFG_CMD_ASKENV		| \
-				 CFG_CMD_DHRYSTONE	| \
+				 CFG_CMD_BOOTD		| \
+				 CFG_CMD_CONSOLE	| \
+				 CFG_CMD_IMI		| \
 				 CFG_CMD_NET    	)
 
 #define CONFIG_COMMANDS_NO	(CFG_CMD_NFS		| \
@@ -80,10 +82,7 @@
 				 CFG_CMD_IMLS		| \
 				 CFG_CMD_ITEST		| \
 				 CFG_CMD_XING		| \
-				 CFG_CMD_IMI		| \
 				 CFG_CMD_BMP		| \
-				 CFG_CMD_BOOTD		| \
-				 CFG_CMD_CONSOLE	| \
 				 CFG_CMD_LOADS		| \
 				 CFG_FLASH_PROTECTION	| \
 				 CFG_CMD_LOADB		)
