Index: linux-2.6.32.42/include/linux/mtd/nand.h
===================================================================
--- linux-2.6.32.42.orig/include/linux/mtd/nand.h	2013-05-08 13:51:18.000000000 +0800
+++ linux-2.6.32.42/include/linux/mtd/nand.h	2013-05-08 14:12:07.000000000 +0800
@@ -183,7 +183,7 @@
 					&& (chip->page_shift > 9))
 
 /* Mask to zero out the chip options, which come from the id table */
-#define NAND_CHIPOPTIONS_MSK	(0x0000ffff & ~NAND_NO_AUTOINCR)
+#define NAND_CHIPOPTIONS_MSK	(0x0000ffff & ~NAND_NO_AUTOINCR & ~NAND_NO_SUBPAGE_WRITE)
 
 /* Non chip related options */
 /* Use a flash based bad block table. This option is passed to the
