--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -686,6 +686,9 @@ static struct flash_info __devinitdata m
 	{ "m25pe80", 0x208014,  0, 64 * 1024, 16, },
 	{ "m25pe16", 0x208015,  0, 64 * 1024, 32, SECT_4K, },
 
+	/* Numonyx (Micron) QUAD-IO */
+	{ "n25q128", 0x20ba18, 0,  64 * 1024, 256, },
+
 	/* Winbond -- w25x "blocks" are 64K, "sectors" are 4KiB */
 	{ "w25x10", 0xef3011, 0, 64 * 1024, 2, SECT_4K, },
 	{ "w25x20", 0xef3012, 0, 64 * 1024, 4, SECT_4K, },
