use single environment variable "ethaddr" for all interfaces

diff --git a/net/eth.c b/net/eth.c
--- a/net/eth.c
+++ b/net/eth.c
@@ -58,7 +58,11 @@ int eth_getenv_enetaddr_by_index(const c
 				 uchar *enetaddr)
 {
 	char enetvar[32];
+#ifndef CONFIG_SINGLE_ETHADDR
 	sprintf(enetvar, index ? "%s%daddr" : "%saddr", base_name, index);
+#else
+	sprintf(enetvar, "%saddr", base_name);
+#endif
 	return eth_getenv_enetaddr(enetvar, enetaddr);
 }
 
